modules/core/core-project/src/main/java/com/enonic/xp/core/impl/project/ProjectNodeEventListener.java

```diff
@@ -0,0 +1,129 @@
+package com.enonic.xp.core.impl.project;
+
+import java.util.List;
+import java.util.Map;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.enonic.xp.content.ContentConstants;
+import com.enonic.xp.content.ContentId;
+import com.enonic.xp.content.ContentService;
+import com.enonic.xp.context.Context;
+import com.enonic.xp.context.ContextBuilder;
+import com.enonic.xp.event.Event;
+import com.enonic.xp.event.EventListener;
+import com.enonic.xp.project.Project;
+import com.enonic.xp.project.ProjectName;
+import com.enonic.xp.project.ProjectService;
+import com.enonic.xp.repository.RepositoryId;
+import com.enonic.xp.security.PrincipalKey;
+import com.enonic.xp.security.RoleKeys;
+import com.enonic.xp.security.User;
+import com.enonic.xp.security.auth.AuthenticationInfo;
+
+@Component(immediate = true)
+public class ProjectNodeEventListener
+    implements EventListener
+{
+    private final static Logger LOG = LoggerFactory.getLogger( ProjectNodeEventListener.class );
+
+    private ProjectService projectService;
+
+    private ContentService contentService;
+
+    @Override
+    public void onEvent( final Event event )
+    {
+        if ( event != null )
+        {
+            createAdminContext().runWith( () -> doHandleEvent( event ) );
+        }
+    }
+
+    private void doHandleEvent( final Event event )
+    {
+        final String type = event.getType();
+
+        switch ( type )
+        {
+            case "node.created":
+                handleEventType( event );
+                break;
+            case "node.updated":
+                handleEventType( event );
+                break;
+        }
+    }
+
+    private void handleEventType( final Event event )
+    {
+        try
+        {
+            final List<Map<String, String>> nodes = (List<Map<String, String>>) event.getData().get( "nodes" );
+
+            nodes.forEach( nodeMap -> {
+                if ( nodeMap.get( "path" ).startsWith( "/content/" ) )
+                {
+                    final ProjectName currentProjectName = ProjectName.from( RepositoryId.from( nodeMap.get( "repo" ) ) );
+
+                    final Project sourceProject = this.projectService.list().
+                        stream().
+                        filter( project -> currentProjectName.equals( project.getName() ) ).
+                        findAny().
+                        orElse( null );
+
+                    this.projectService.list().
+                        stream().
+                        filter( project -> currentProjectName.equals( project.getParent() ) ).
+                        forEach( targetProject -> {
```

Comment:
Offload loops and heavy tasks into a separate thread (it still need to be in order, so for now single threaded thread pool would work)